src/server.c

```diff
@@ -135,24 +135,43 @@ void get_file(int fd, struct cache *cache, char *request_path)
     struct file_data *filedata; 
     char *mime_type;
 
-    // Create filepath and load file data.
+
+    // Create filepath.
     snprintf(filepath, sizeof filepath, "%s%s", SERVER_ROOT, request_path);
-    filedata = file_load(filepath);
 
-    // Check if file data is null and throw 404.
-    if (filedata == NULL) {
-        resp_404(fd);
-        exit(3);
-    }
+    // First check to see if the path to the file is in the cache (use the file path as the key).
+    struct cache_entry *entry = cache_get(cache, filepath);
+
+    // If it's there, serve it back.
+    if (entry != NULL) {
+        // Create send response.
+        send_response(fd, "HTTP/1.1 200 OK", entry->content_type, entry->content, entry->content_length);
     
-    // Get mime type for file path.
-    mime_type = mime_type_get(filepath);
+    // If it's not there...
+    } else {
+        // Load file data from disk.
+        filedata = file_load(filepath);
 
-    // Create send response. 
-    send_response(fd, "HTTP/1.1 200 OK", mime_type, filedata->data, filedata->size);
+        // Check if file data is null and throw 404.
+        if (filedata == NULL) {
+            resp_404(fd);
+            exit(3);
```

Comment:
having a web server exit when hitting a 404 isn't the best plan :)
Go with `return;` instead.